File: packet_seaquence/log/latest/work/src/protocol_analyzer/arp.py
#!/usr/bin/env python3
"""
ARPプロトコル解析クラスを提供するモジュール
"""
from src.protocol_analyzer.base import ProtocolAnalyzer


class AnalyzeArp(ProtocolAnalyzer):
    """
    ARP内の詳細情報を解析するクラス
    """
    
    def analyze(self):
        """
        パケットからARP情報を解析する
        
        Returns:
            dict: ARP情報を含む辞書、または情報がない場合はNone
        """
        if not hasattr(self.packet, 'arp'):
            return None
        
        try:
            arp_info = {}
            
            # ARPオペレーションコード
            if hasattr(self.packet.arp, 'opcode'):
                arp_code = int(self.packet.arp.opcode)
                arp_info['operation_code'] = arp_code
                arp_info['operation'] = self._get_arp_operation(arp_code)
            elif hasattr(self.packet.arp, 'operation'):
                arp_code = int(self.packet.arp.operation)
                arp_info['operation_code'] = arp_code
                arp_info['operation'] = self._get_arp_operation(arp_code)
            
            # 送信元MACアドレス
            if hasattr(self.packet.arp, 'src_hw_mac'):
                arp_info['src_mac'] = self.packet.arp.src_hw_mac
            elif hasattr(self.packet.arp, 'src.hw_mac'):
                arp_info['src_mac'] = self.packet.arp.src.hw_mac
            
            # 送信元IPアドレス
            if hasattr(self.packet.arp, 'src_proto_ipv4'):
                arp_info['src_ip'] = self.packet.arp.src_proto_ipv4
            elif hasattr(self.packet.arp, 'src.proto_ipv4'):
                arp_info['src_ip'] = self.packet.arp.src.proto_ipv4
            
            # 宛先MACアドレス
            if hasattr(self.packet.arp, 'dst_hw_mac'):
                arp_info['dst_mac'] = self.packet.arp.dst_hw_mac
            elif hasattr(self.packet.arp, 'dst.hw_mac'):
                arp_info['dst_mac'] = self.packet.arp.dst.hw_mac
            
            # 宛先IPアドレス
            if hasattr(self.packet.arp, 'dst_proto_ipv4'):
                arp_info['dst_ip'] = self.packet.arp.dst_proto_ipv4
            elif hasattr(self.packet.arp, 'dst.proto_ipv4'):
                arp_info['dst_ip'] = self.packet.arp.dst.proto_ipv4
            
            return arp_info
            
        except Exception as e:
            print(f"ARP解析中にエラーが発生しました: {str(e)}")
            return None
    
    def get_display_info(self):
        """
        表示用のARP情報を取得する
        
        Returns:
            str: 表示用のARP情報
        """
        arp_info = self.analyze()
        if not arp_info:
            return "ARP情報なし"
        
        operation = arp_info.get('operation', '不明')
        src_mac = arp_info.get('src_mac', '不明')
        src_ip = arp_info.get('src_ip', '不明')
        dst_mac = arp_info.get('dst_mac', '不明')
        dst_ip = arp_info.get('dst_ip', '不明')
        
        if 'REQUEST' in operation:
            return f"ARP Request: {src_ip} ({src_mac}) は {dst_ip} のMACアドレスを要求"
        elif 'REPLY' in operation:
            return f"ARP Reply: {src_ip} は {src_mac} です"
        else:
            return f"ARP {operation}: {src_ip} ({src_mac}) -> {dst_ip} ({dst_mac})"
    
    def get_summary(self):
        """
        ARP情報のサマリーを取得する
        
        Returns:
            str: ARP情報のサマリー
        """
        arp_info = self.analyze()
        if not arp_info:
            return "ARP"
        
        operation = arp_info.get('operation', '').split('-')[0]  # REQUEST, REPLY などの部分だけ取得
        src_ip = arp_info.get('src_ip', '')
        dst_ip = arp_info.get('dst_ip', '')
        
        return f"ARP {operation} {src_ip} -> {dst_ip}"
    
    def _get_arp_operation(self, code):
        """
        ARPオペレーションコードの説明を取得する
        
        Args:
            code (int): ARPオペレーションコード
        
        Returns:
            str: ARPオペレーションの説明
        """
        arp_operations = {
            1: 'REQUEST',
            2: 'REPLY',
            3: 'RARP-REQUEST',
            4: 'RARP-REPLY',
            5: 'DRARP-REQUEST',
            6: 'DRARP-REPLY',
            7: 'DRARP-ERROR',
            8: 'InARP-REQUEST',
            9: 'InARP-REPLY'
        }
        return arp_operations.get(code, f'UNKNOWN ({code})')

File: packet_seaquence/log/go_src/protocol_analyzer/arp.go
// Package protocol_analyzer provides ARP protocol analyzer
package protocol_analyzer

import (
	"fmt"
	"strconv"
	"strings"
)

// AnalyzeARP implements the ProtocolAnalyzer interface for ARP
type AnalyzeARP struct {
	Pkt Packet
}

// NewAnalyzeARP creates a new ARP analyzer
func NewAnalyzeARP(packet Packet) *AnalyzeARP {
	return &AnalyzeARP{
		Pkt: packet,
	}
}

// Analyze extracts ARP information from a packet
func (a *AnalyzeARP) Analyze() map[string]interface{} {
	if !a.Pkt.LayerExists("arp") {
		return nil
	}

	arpInfo := make(map[string]interface{})
	
	// ARP operation code
	if opcodeVal, ok := a.Pkt.GetField("arp", "opcode"); ok {
		opcode, err := strconv.Atoi(fmt.Sprintf("%v", opcodeVal))
		if err == nil {
			arpInfo["operation_code"] = opcode
			arpInfo["operation"] = a.getARPOperation(opcode)
		}
	} else if operationVal, ok := a.Pkt.GetField("arp", "operation"); ok {
		operation, err := strconv.Atoi(fmt.Sprintf("%v", operationVal))
		if err == nil {
			arpInfo["operation_code"] = operation
			arpInfo["operation"] = a.getARPOperation(operation)
		}
	}
	
	// Source MAC address
	if srcMac, ok := a.Pkt.GetField("arp", "src_hw_mac"); ok {
		arpInfo["src_mac"] = srcMac
	} else if srcMac, ok := a.Pkt.GetField("arp", "src.hw_mac"); ok {
		arpInfo["src_mac"] = srcMac
	}
	
	// Source IP address
	if srcIP, ok := a.Pkt.GetField("arp", "src_proto_ipv4"); ok {
		arpInfo["src_ip"] = srcIP
	} else if srcIP, ok := a.Pkt.GetField("arp", "src.proto_ipv4"); ok {
		arpInfo["src_ip"] = srcIP
	}
	
	// Destination MAC address
	if dstMac, ok := a.Pkt.GetField("arp", "dst_hw_mac"); ok {
		arpInfo["dst_mac"] = dstMac
	} else if dstMac, ok := a.Pkt.GetField("arp", "dst.hw_mac"); ok {
		arpInfo["dst_mac"] = dstMac
	}
	
	// Destination IP address
	if dstIP, ok := a.Pkt.GetField("arp", "dst_proto_ipv4"); ok {
		arpInfo["dst_ip"] = dstIP
	} else if dstIP, ok := a.Pkt.GetField("arp", "dst.proto_ipv4"); ok {
		arpInfo["dst_ip"] = dstIP
	}
	
	return arpInfo
}

// GetDisplayInfo returns formatted ARP information
func (a *AnalyzeARP) GetDisplayInfo() string {
	info := a.Analyze()
	if info == nil {
		return "ARP情報なし"
	}
	
	operation, _ := info["operation"].(string)
	if operation == "" {
		operation = "不明"
	}
	
	srcMac, _ := info["src_mac"].(string)
	if srcMac == "" {
		srcMac = "不明"
	}
	
	srcIP, _ := info["src_ip"].(string)
	if srcIP == "" {
		srcIP = "不明"
	}
	
	dstMac, _ := info["dst_mac"].(string)
	if dstMac == "" {
		dstMac = "不明"
	}
	
	dstIP, _ := info["dst_ip"].(string)
	if dstIP == "" {
		dstIP = "不明"
	}
	
	if operation == "REQUEST" {
		return fmt.Sprintf("ARP Request: %s (%s) は %s のMACアドレスを要求", srcIP, srcMac, dstIP)
	} else if operation == "REPLY" {
		return fmt.Sprintf("ARP Reply: %s は %s です", srcIP, srcMac)
	} else {
		return fmt.Sprintf("ARP %s: %s (%s) -> %s (%s)", operation, srcIP, srcMac, dstIP, dstMac)
	}
}

// GetSummary returns a summary of ARP information
func (a *AnalyzeARP) GetSummary() string {
	info := a.Analyze()
	if info == nil {
		return "ARP"
	}
	
	var operation string
	if op, ok := info["operation"].(string); ok {
		// Split at hyphen and get first part (e.g., "REPLY-REQUEST" -> "REPLY")
		parts := strings.Split(op, "-")
		operation = parts[0]
	} else {
		operation = ""
	}
	
	srcIP, _ := info["src_ip"].(string)
	if srcIP == "" {
		srcIP = "?"
	}
	
	dstIP, _ := info["dst_ip"].(string)
	if dstIP == "" {
		dstIP = "?"
	}
	
	return fmt.Sprintf("ARP %s %s -> %s", operation, srcIP, dstIP)
}

// getARPOperation returns a human-readable ARP operation string
func (a *AnalyzeARP) getARPOperation(code int) string {
	arpOperations := map[int]string{
		1: "REQUEST",
		2: "REPLY",
		3: "RARP-REQUEST",
		4: "RARP-REPLY",
		5: "DRARP-REQUEST",
		6: "DRARP-REPLY",
		7: "DRARP-ERROR",
		8: "InARP-REQUEST",
		9: "InARP-REPLY",
	}
	
	if op, ok := arpOperations[code]; ok {
		return op
	}
	return fmt.Sprintf("UNKNOWN (%d)", code)
}